### ISSUE TITLE
In the function backward_D_basic, why operate detach() for fake???

### ISSUE THREAD

**Issue by astringentEG** [junyanz/pytorch-CycleGAN-and-pix2pix#388]
def backward_D_basic(self, netD, real, fake):
        # Real
        pred_real = netD(real)
        loss_D_real = self.criterionGAN(pred_real, True)
        # Fake
        pred_fake = netD(fake.detach())
        loss_D_fake = self.criterionGAN(pred_fake, False)
        # Combined loss
        loss_D = (loss_D_real + loss_D_fake) * 0.5
        # backward
        loss_D.backward()
        return loss_D

Q1. What is detach() in pytorch?
      I know this is very basic question, but I cannot understand the explanation in document.
      (https://pytorch.org/docs/stable/tensors.html)
      It seems detach() prevent something related with graident & backprop...

Q2. Why do detach() for fake only?
      I thought D care about real and fake equally.
      Actually, because I don't know what is role of detach(), so I also cannot know it.

I'm very new in this field, so kind explanation will be greatly appreciated.
Thank you.

**Comment by junyanz** [comment]
detach() stops the gradient. fake.detach() will make sure that G does not get gradients. Real does not need to be detached as real is a constant, not a variable.
